### ISSUE TITLE
iOS recording fails second time

### ISSUE THREAD

**Issue by igorosabel** [chris-rudmin/opus-recorder#158]
Hi!

I already saw an issue regarding this topic ([this](https://github.com/chris-rudmin/opus-recorder/issues/141)) , but the same thing happens to me.

I've tried the example you gave there on an iPhone 6S with iOS 11.4 and it works... so I don't know where my problem is coming from.

I'm using a different approach and I'd like some help as I've been stuck for days.

I ask for the navigator.getUserMedia as soon as the user comes into the recording page, to test if the user has microphone, to present him with a message explaining the permission dialog... And when the user agrees with the permission I show him the recorder interface. The flow is something like this:

I initialize the recorder:
```
var recorder = new Recorder({
  encoderPath: '/js/lib/encoderWorker.min.js',
  leaveStreamOpen: true // as the user can make several recordings until he is happy with it
});
recorder.ondataavailable = uploadRecording; // A function just to upload the recording
```
Then I ask for the getUserMedia:
```
navigator.mediaDevices.getUserMedia({ audio: true })
.then(onSuccess)
.catch(onError);
```

The onsuccess function does this:
```
function onSuccess(stream) {
  var AudioContext = window.AudioContext || window.webkitAudioContext;
  var audioCtx = new AudioContext();
  sourceNode = audioCtx.createMediaStreamSource( stream ); // sourceNode is a global variable
  recorder.stream = stream;
  // ... show the recorder interface
}
```
The record button launches this function:
```
function startRecording(){
  numRecordings++; // numRecordings is previously initialized to 0
  if (numRecordings==1){
    // The first time I pass the sourceNode variable as it is not initialized
    recorder.start(sourceNode).catch(function(e){ console.log(e); });
  }
  else{
    // Next times I don't pass the sourceNode since it is already initialized
    recorder.start().catch(function(e){ console.log(e); });
  }
}
```

The stop button basically does this:
```
function stopRecording(){
  recorder.stop();
}
```

It works perfectly on Firefox (win and mac), Chrome (win and mac), Edge (win10). But on Safari (both on iOS and macOS) the second, third and so on recordings are just blank recordings.

Any help? Maybe I should do things in another way?

Thank you very much, I think it is a great library :)

Iñigo

**Comment by chris-rudmin** [comment]
The recorder will ask for mic permissions on `start`. Also you can pass the source node to `start()` instead of assigning it to the recorder instance.

**Comment by igorosabel** [comment]
Hi, thanks for the fast response.

`The recorder will ask for mic permissions on start`

That's precissely what I am trying to avoid. When the user goes into the page, there is a message explaining that it has to click on "allow", a screenshot of the permission dialog... And when the user allows microphone access, I hide the message and show the recorder.

So asking for permission on start is no good for me.

`Also you can pass the source node to start() instead of assigning it to the recorder instance.`

That is what I do, on the first recording click I pass the sourceNode variable to the start method, and on the next calls I just call start since it is already initialized.

Thanks again.

**Comment by chris-rudmin** [comment]
Safari is a bit more restrictive on permissions. The reason that permissions are asked on `start` is because the use of mic stream needs to be initiated by a user action (button click). Try using the recorder with it's default behaviour, and if it works then it would point to an issue with your implementation and accessing the mic stream in Safari.

**Comment by igorosabel** [comment]
Ok, I have done a couple of minimal tests:

[Ask for microphone on page load](https://osumi.es/recorder/test.html)

[Ask for microphone on record start (your example)](https://osumi.es/recorder/test2.html)

With the first test, there is no error, but all the recordings I do are blank. The recording duration is correct (in seconds) but no sound at all.

On the second test, your approach, the first recording is done correctly, but second, third... and so on, are also blank.

I have looked up the blank files with ffprobe, ffmpeg's test tool, and it says this:

```
Input #0, ogg, from 'ios-2.ogg':
Duration: 00:00:04.34, start: 0.000000, bitrate: 1 kb/s
Stream #0:0: Audio: opus, 48000 Hz, mono, fltp
```

These tests were done on an iPhone 6S using Safari on iOS 11.4.

Thanks.

**Comment by chris-rudmin** [comment]
@igorosabel Thanks for the investigation. I was able to replicate your issue, and it seems to be caused by the `leaveStreamOpen` option.  I am leaning towards killing this option.  In all browsers except firefox, if you received permissions once for the mic, it does not ask again. What do you think?

**Comment by igorosabel** [comment]
Nope, that's not the problem :( In both my tests I set the `leaveStreamOpen` to true, so the `clearStream` function is never called.

I think the issue comes from those connect/disconnect calls, I'm testing and I'll tell you if I find something.

Thanks for the help though.

**Comment by chris-rudmin** [comment]
@igorosabel The [included example](https://rawgit.com/chris-rudmin/opus-recorder/master/example/encoder.html) is able to record multiple times on iOS.  The only difference I see is the leaveStreamOpen option. Perhaps I missed something?
